TickSpec/TickSpec.fs

```diff
@@ -98,69 +111,103 @@ type StepDefinitions (givens,whens,thens,events,valueParsers) =
-                        features@features' |> List.filter (not << String.IsNullOrEmpty),
-                            scenarios@scenarios' |> List.filter (not << String.IsNullOrEmpty),
-                                m
+            |> Seq.collect (fun t -> t.GetMethods())
+        StepDefinitions(methods)
+    internal new (methods:MethodInfo seq) =
+        let categorizedMethods =
+            let getScope attributes =
+                attributes
+                |> Seq.cast
+                |> Seq.fold (fun (tags,features,scenarios) (x:StepScopeAttribute) ->
+                    x.Tag::tags, x.Feature::features, x.Scenario::scenarios
+                ) ([],[],[])
+            
+            let attributeMap = Dictionary<Type, (MethodWithScope * obj) list>()
+            let parentScope = Dictionary<Type, MethodScope>()
+            let methodScope = Dictionary<MethodInfo, MethodScope>()
+
+            let attributes = [|
+                typeof<GivenAttribute>; typeof<WhenAttribute>; typeof<ThenAttribute>
+                typeof<BeforeScenarioAttribute>; typeof<AfterScenarioAttribute>
+                typeof<BeforeStepAttribute>; typeof<AfterStepAttribute>
+                typeof<ParserAttribute>
+            |]
+
+            // Initialize the attribute map
+            attributes |> Array.iter (fun attrType -> attributeMap.[attrType] <- List.empty)
+
+            // Iterate through all methods
+            methods |> Seq.iter (fun mi ->
+                // Get all attributes of the method
+                mi.GetCustomAttributes(true)
+                |> Array.iter (fun attr ->
+                    let usedType = attr.GetType()
+                    let correspondingAttrType =
+                        attributes
+                        |> Array.tryFind (fun x -> x.IsAssignableFrom(usedType))
+
+                    match correspondingAttrType with
+                    // In case it is one of the ones we care about, we add it to the map
+                    | Some attrType ->
+                        // We use caching to not repeatedly get the scope for methods or its declaring types
+                        if methodScope.ContainsKey mi |> not then
+                            let parentType = mi.DeclaringType
+                            if parentScope.ContainsKey parentType |> not then
+                                let parentScopeAttribute = parentType.GetCustomAttributes(typeof<StepScopeAttribute>,true)
+                                parentScope.[parentType] <- getScope parentScopeAttribute
+                            let tags', features', scenarios' = parentScope.[parentType]
+                            let methodScopeAttribute = mi.GetCustomAttributes(typeof<StepScopeAttribute>,true)
+                            let tags, features, scenarios = getScope methodScopeAttribute
+                            methodScope.[mi] <- (
+                                tags@tags' |> List.filter (not << String.IsNullOrEmpty),
```

Comment:
@ is very inefficient, as is rewriting with filtering, so the filtering moves into the gather
if these are coming back as arrays, Array.append is much better
(for ultimate efficiency you would pass in the accumulator into getScope but most of the win is already here and its clearer this way)

Reply:
I was thinking about changing it to arrays, but that would require bigger changes (or just changing them to list back). Considering how many values I expect here, the list should be absolutely fine and if not, then we should use set instead of array anyway. So, I will keep it to use the fold but will use the idea of accumulator and will move the filtering in the getScope method.